File: lib/core/repositories/modules_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../models/module_model.dart';
import '../supabase/supabase_client.dart';

final modulesProvider = FutureProvider<List<ModuleModel>>((ref) async {
  final data = await supabase.from('modules').select().order('created_at');
  return (data as List).map((e) => ModuleModel.fromJson(e)).toList();
});

final moduleBySlugProvider =
    FutureProvider.family<ModuleModel?, String>((ref, slug) async {
  final data = await supabase
      .from('modules')
      .select()
      .eq('slug', slug)
      .maybeSingle();
  if (data == null) return null;
  return ModuleModel.fromJson(data);
});

class ModulesRepository {
  Future<ModuleModel> upsert(ModuleModel module) async {
    final payload = {...module.toJson(), 'id': module.id};
    final data = await supabase
        .from('modules')
        .upsert(payload)
        .select()
        .single();
    return ModuleModel.fromJson(data);
  }

  Future<void> delete(String id) async {
    await supabase.from('modules').delete().eq('id', id);
  }
}

final modulesRepositoryProvider =
    Provider<ModulesRepository>((ref) => ModulesRepository());

File: src/core/repositories/modules.repository.ts
import { supabase } from "@/core/supabase/client";
import { moduleFromRow, type ModuleModel } from "@/core/models/module";

export async function fetchModules(): Promise<ModuleModel[]> {
  const { data, error } = await supabase
    .from("modules")
    .select("*")
    .order("created_at");
  if (error) throw error;
  return data.map(moduleFromRow);
}

export async function fetchModuleBySlug(slug: string): Promise<ModuleModel | null> {
  const { data, error } = await supabase
    .from("modules")
    .select("*")
    .eq("slug", slug)
    .maybeSingle();
  if (error) throw error;
  return data ? moduleFromRow(data) : null;
}

export async function fetchModulesBySlugs(slugs: string[]): Promise<ModuleModel[]> {
  if (slugs.length === 0) return [];
  const { data, error } = await supabase
    .from("modules")
    .select("*")
    .in("slug", slugs);
  if (error) throw error;
  return data.map(moduleFromRow);
}

export async function upsertModule(
  slug: string,
  content: string
): Promise<void> {
  const { error } = await supabase
    .from("modules")
    .update({ content })
    .eq("slug", slug);
  if (error) throw error;
}
